File: apps/api/src/smplat_api/models/product.py
from enum import Enum
from uuid import uuid4

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    Enum as SqlEnum,
    ForeignKey,
    Integer,
    JSON,
    Numeric,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from smplat_api.db.base import Base
from .customer_profile import CurrencyEnum


class ProductStatusEnum(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    ARCHIVED = "archived"


class ProductOptionGroupTypeEnum(str, Enum):
    SINGLE = "single"
    MULTIPLE = "multiple"


class ProductCustomFieldTypeEnum(str, Enum):
    TEXT = "text"
    URL = "url"
    NUMBER = "number"


class ProductSubscriptionBillingCycleEnum(str, Enum):
    ONE_TIME = "one_time"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    ANNUAL = "annual"


class Product(Base):
    __tablename__ = "products"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    slug = Column(String, nullable=False, unique=True)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    category = Column(String, nullable=False)
    base_price = Column(Numeric(10, 2), nullable=False)
    currency = Column(
        SqlEnum(CurrencyEnum, name="preferred_currency_enum", create_type=False),
        nullable=False,
        server_default=CurrencyEnum.EUR.value,
    )
    status = Column(
        SqlEnum(ProductStatusEnum, name="product_status_enum"),
        nullable=False,
        server_default=ProductStatusEnum.DRAFT.value,
    )
    channel_eligibility = Column(JSON, nullable=False, default=list)
    fulfillment_config = Column(JSON, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    option_groups = relationship(
        "ProductOptionGroup",
        back_populates="product",
        order_by="ProductOptionGroup.display_order",
        cascade="all, delete-orphan",
    )
    media_assets = relationship(
        "ProductMediaAsset",
        back_populates="product",
        order_by="ProductMediaAsset.created_at.desc()",
        cascade="all, delete-orphan",
    )
    audit_logs = relationship(
        "ProductAuditLog",
        back_populates="product",
        order_by="ProductAuditLog.created_at.desc()",
        cascade="all, delete-orphan",
    )
    add_ons = relationship(
        "ProductAddOn",
        back_populates="product",
        order_by="ProductAddOn.display_order",
        cascade="all, delete-orphan",
    )
    custom_fields = relationship(
        "ProductCustomField",
        back_populates="product",
        order_by="ProductCustomField.display_order",
        cascade="all, delete-orphan",
    )
    subscription_plans = relationship(
        "ProductSubscriptionPlan",
        back_populates="product",
        order_by="ProductSubscriptionPlan.display_order",
        cascade="all, delete-orphan",
    )

    @property
    def fulfillment_summary(self) -> dict | None:
        """Return structured fulfillment assurances derived from configuration."""

        config = self.fulfillment_config or {}
        if not isinstance(config, dict):
            return None

        delivery_config = config.get("delivery") or {}
        assurances_config = config.get("assurances") or []
        support_config = config.get("support") or []

        def _as_int(value: object) -> int | None:
            try:
                if isinstance(value, bool):
                    return int(value)
                if isinstance(value, (int, float)):
                    return int(value)
                if isinstance(value, str) and value.strip():
                    return int(float(value))
            except Exception:  # pragma: no cover - defensive conversion
                return None
            return None

        def _as_str(value: object) -> str | None:
            if isinstance(value, str) and value.strip():
                return value.strip()
            return None

        def _extract_delivery(raw: dict | None) -> dict | None:
            if not isinstance(raw, dict):
                return None

            return {
                "minDays": _as_int(raw.get("minDays") or raw.get("min_days")),
                "maxDays": _as_int(raw.get("maxDays") or raw.get("max_days")),
                "averageDays": _as_int(raw.get("averageDays") or raw.get("average_days")),
                "confidence": _as_str(raw.get("confidence")),
                "headline": _as_str(raw.get("headline")),
                "narrative": _as_str(raw.get("narrative")),
            }

        def _extract_assurances(raw_list: list | None) -> list[dict]:
            items: list[dict] = []
            if not isinstance(raw_list, list):
                return items
            for entry in raw_list:
                if not isinstance(entry, dict):
                    continue
                label = _as_str(entry.get("label"))
                description = _as_str(entry.get("description"))
                if not label and not description:
                    continue
                items.append(
                    {
                        "id": _as_str(entry.get("id")) or label or "assurance",
                        "label": label or description or "",
                        "description": description,
                        "evidence": _as_str(entry.get("evidence")),
                        "source": _as_str(entry.get("source")),
                    }
                )
            return items

        def _extract_support(raw_list: list | None) -> list[dict]:
            items: list[dict] = []
            if not isinstance(raw_list, list):
                return items
            for entry in raw_list:
                if not isinstance(entry, dict):
                    continue
                channel = _as_str(entry.get("channel")) or _as_str(entry.get("type"))
                label = _as_str(entry.get("label")) or channel
                target = _as_str(entry.get("target")) or _as_str(entry.get("href"))
                if not channel or not label or not target:
                    continue
                items.append(
                    {
                        "id": _as_str(entry.get("id")) or f"{channel}:{target}",
                        "channel": channel,
                        "label": label,
                        "target": target,
                        "availability": _as_str(entry.get("availability")),
                    }
                )
            return items

        summary: dict[str, object] = {}
        delivery = _extract_delivery(delivery_config)
        assurances = _extract_assurances(assurances_config)
        support = _extract_support(support_config)

        if delivery and any(value is not None for value in delivery.values()):
            summary["delivery"] = delivery
        if assurances:
            summary["assurances"] = assurances
        if support:
            summary["support"] = support

        return summary or None


class ProductOptionGroup(Base):
    __tablename__ = "product_option_groups"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    name = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    group_type = Column(
        SqlEnum(ProductOptionGroupTypeEnum, name="product_option_group_type_enum"),
        nullable=False,
        server_default=ProductOptionGroupTypeEnum.SINGLE.value,
    )
    is_required = Column(Boolean, nullable=False, server_default="false")
    display_order = Column(Integer, nullable=False, server_default="0")
    metadata_json = Column(JSON, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    product = relationship("Product", back_populates="option_groups")
    options = relationship(
        "ProductOption",
        back_populates="group",
        order_by="ProductOption.display_order",
        cascade="all, delete-orphan",
    )


class ProductOption(Base):
    __tablename__ = "product_options"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    group_id = Column(UUID(as_uuid=True), ForeignKey("product_option_groups.id", ondelete="CASCADE"), nullable=False)
    name = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    price_delta = Column(Numeric(10, 2), nullable=False, server_default="0")
    metadata_json = Column(JSON, nullable=True)
    display_order = Column(Integer, nullable=False, server_default="0")
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    group = relationship("ProductOptionGroup", back_populates="options")


class ProductAddOn(Base):
    __tablename__ = "product_add_ons"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    label = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    price_delta = Column(Numeric(10, 2), nullable=False, server_default="0")
    is_recommended = Column(Boolean, nullable=False, server_default="false")
    display_order = Column(Integer, nullable=False, server_default="0")
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    product = relationship("Product", back_populates="add_ons")


class ProductCustomField(Base):
    __tablename__ = "product_custom_fields"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    label = Column(String, nullable=False)
    field_type = Column(
        SqlEnum(ProductCustomFieldTypeEnum, name="product_custom_field_type_enum"),
        nullable=False,
        server_default=ProductCustomFieldTypeEnum.TEXT.value,
    )
    placeholder = Column(String, nullable=True)
    help_text = Column(Text, nullable=True)
    is_required = Column(Boolean, nullable=False, server_default="false")
    display_order = Column(Integer, nullable=False, server_default="0")
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    product = relationship("Product", back_populates="custom_fields")


class ProductSubscriptionPlan(Base):
    __tablename__ = "product_subscription_plans"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    label = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    billing_cycle = Column(
        SqlEnum(ProductSubscriptionBillingCycleEnum, name="product_subscription_billing_cycle_enum"),
        nullable=False,
        server_default=ProductSubscriptionBillingCycleEnum.ONE_TIME.value,
    )
    price_multiplier = Column(Numeric(10, 2), nullable=True)
    price_delta = Column(Numeric(10, 2), nullable=True)
    is_default = Column(Boolean, nullable=False, server_default="false")
    display_order = Column(Integer, nullable=False, server_default="0")
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    product = relationship("Product", back_populates="subscription_plans")


class ProductMediaAsset(Base):
    """Uploaded asset metadata associated with a product."""

    # meta: provenance: product-merchandising
    __tablename__ = "product_media_assets"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    label = Column(String(150), nullable=True)
    asset_url = Column(String(1024), nullable=False)
    storage_key = Column(String(512), nullable=True)
    metadata_json = Column("metadata", JSON, nullable=False, default=dict)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    product = relationship("Product", back_populates="media_assets")


class ProductAuditLog(Base):
    """Immutable audit log capturing product merchandising changes."""

    # meta: provenance: product-merchandising
    __tablename__ = "product_audit_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id", ondelete="SET NULL"), nullable=True)
    action = Column(String, nullable=False)
    actor_user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    actor_email = Column(String(255), nullable=True)
    before_snapshot = Column(JSON, nullable=True)
    after_snapshot = Column(JSON, nullable=True)
    metadata_json = Column("metadata", JSON, nullable=False, default=dict)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    product = relationship("Product", back_populates="audit_logs")

File: apps/web/src/types/product.ts
import type {
  ServiceOverrideCondition as SharedServiceOverrideCondition,
  ServiceOverrideRule as SharedServiceOverrideRule,
} from "@smplat/types";

export type ServiceOverrideCondition = SharedServiceOverrideCondition;
export type ServiceOverrideRule = SharedServiceOverrideRule;

export type ProductOptionDiscountTier = {
  minAmount: number;
  unitPrice: number;
  label?: string | null;
};

export type ProductOptionStructuredPricing = {
  amount: number;
  amountUnit: string;
  basePrice: number;
  unitPrice: number;
  dripMinPerDay?: number | null;
  discountTiers?: ProductOptionDiscountTier[] | null;
};

export type ProductAssetUsageTag = "hero" | "detail" | "gallery" | "social" | (string & {});

export type ProductOptionMediaAttachment = {
  assetId: string;
  clientId?: string | null;
  displayOrder?: number | null;
  isPrimary?: boolean | null;
  usage?: string | null;
  usageTags?: ProductAssetUsageTag[] | null;
  label?: string | null;
  altText?: string | null;
  storageKey?: string | null;
  checksum?: string | null;
};

export type ProductOptionCalculatorMetadata = {
  expression: string;
  sampleAmount?: number | null;
  sampleDays?: number | null;
};

export type ProductOptionMetadata = {
  structuredPricing?: ProductOptionStructuredPricing | null;
  media?: ProductOptionMediaAttachment[] | null;
  recommended?: boolean;
  editorKey?: string;
  marketingTagline?: string | null;
  fulfillmentSla?: string | null;
  heroImageUrl?: string | null;
  calculator?: ProductOptionCalculatorMetadata | null;
} & Record<string, unknown>;

export type ServiceOverrideConfig = {
  providerId?: string | null;
  costAmount?: number | null;
  costCurrency?: string | null;
  marginTarget?: number | null;
  fulfillmentMode?: "immediate" | "scheduled" | "refill";
  payloadTemplate?: Record<string, unknown> | null;
  dripPerDay?: number | null;
  previewQuantity?: number | null;
  rules?: ServiceOverrideRule[] | null;
};

export type ProductAddOnPricing =
  | { mode: "flat"; amount: number }
  | { mode: "percentage"; amount: number }
  | ({ mode: "serviceOverride"; serviceId: string; amount?: number | null } & ServiceOverrideConfig);

export type ProductAddOnMetadata = {
  pricing?: ProductAddOnPricing | null;
  editorKey?: string;
} & Record<string, unknown>;

export type ProductAddOnPricingSnapshot = {
  mode: "flat" | "percentage" | "serviceOverride";
  amount?: number | null;
  percentageMultiplier?: number | null;
  serviceId?: string | null;
  serviceAction?: string | null;
  serviceProviderId?: string | null;
  serviceProviderName?: string | null;
  serviceDescriptor?: Record<string, unknown> | null;
  providerCostAmount?: number | null;
  providerCostCurrency?: string | null;
  marginTarget?: number | null;
  fulfillmentMode?: "immediate" | "scheduled" | "refill";
  payloadTemplate?: Record<string, unknown> | null;
  dripPerDay?: number | null;
  previewQuantity?: number | null;
  serviceRules?: ServiceOverrideRule[] | null;
};

export type ProductOption = {
  id: string;
  label: string;
  description?: string | null;
  priceDelta: number;
  metadataJson?: ProductOptionMetadata | null;
  displayOrder: number;
};

export type ProductOptionGroup = {
  id: string;
  name: string;
  description?: string | null;
  groupType: "single" | "multiple";
  isRequired: boolean;
  displayOrder: number;
  metadataJson?: Record<string, unknown> | null;
  options: ProductOption[];
};

export type ProductAddOn = {
  id: string;
  label: string;
  description?: string | null;
  priceDelta: number;
  isRecommended: boolean;
  displayOrder: number;
  metadataJson?: ProductAddOnMetadata | null;
  pricing?: ProductAddOnPricingSnapshot | null;
  computedDelta: number;
  percentageMultiplier?: number | null;
};

export type CustomFieldVisibilityCondition =
  | { kind: "option"; optionId?: string; optionKey?: string; groupId?: string; groupKey?: string }
  | { kind: "addOn"; addOnId?: string; addOnKey?: string }
  | { kind: "subscriptionPlan"; planId?: string; planKey?: string }
  | { kind: "channel"; channel: string };

export type ProductCustomFieldRegexRule = {
  pattern: string;
  flags?: string;
  description?: string | null;
  sampleValue?: string | null;
};

export type ProductCustomFieldValidationRules = {
  minLength?: number;
  maxLength?: number;
  minValue?: number;
  maxValue?: number;
  pattern?: string;
  regex?: ProductCustomFieldRegexRule | null;
  disallowWhitespace?: boolean;
  numericStep?: number;
  allowedValues?: string[];
};

export type ProductCustomFieldVisibilityRules = {
  mode: "all" | "any";
  conditions: CustomFieldVisibilityCondition[];
};

export type ProductCustomFieldMetadata = {
  helperText?: string | null;
  sampleValues?: string[] | null;
  validationRules?: ProductCustomFieldValidationRules | null;
  validation?: ProductCustomFieldValidationRules | null;
  defaultValue?: string | null;
  passthrough?: {
    checkout?: boolean;
    fulfillment?: boolean;
  } | null;
  visibilityRules?: ProductCustomFieldVisibilityRules | null;
  conditionalVisibility?: ProductCustomFieldVisibilityRules | null;
  regexTester?: {
    sampleValue?: string | null;
    lastResult?: boolean | null;
  } | null;
} & Record<string, unknown>;

export type ProductCustomField = {
  id: string;
  label: string;
  fieldType: "text" | "url" | "number";
  placeholder?: string | null;
  helpText?: string | null;
  isRequired: boolean;
  displayOrder: number;
  metadataJson?: ProductCustomFieldMetadata | null;
  validationRules?: ProductCustomFieldValidationRules | null;
  validation?: ProductCustomFieldValidationRules | null;
  defaultValue?: string | null;
  visibilityRules?: ProductCustomFieldVisibilityRules | null;
  conditionalVisibility?: ProductCustomFieldMetadata["conditionalVisibility"];
  passthroughTargets?: NonNullable<ProductCustomFieldMetadata["passthrough"]>;
};

export type ProductSubscriptionPlan = {
  id: string;
  label: string;
  description?: string | null;
  billingCycle: "one_time" | "monthly" | "quarterly" | "annual";
  priceMultiplier?: number | null;
  priceDelta?: number | null;
  isDefault: boolean;
  displayOrder: number;
};

export type ProductDeliveryEstimate = {
  minDays?: number | null;
  maxDays?: number | null;
  averageDays?: number | null;
  confidence?: string | null;
  headline?: string | null;
  narrative?: string | null;
};

export type ProductAssurancePoint = {
  id: string;
  label: string;
  description?: string | null;
  evidence?: string | null;
  source?: string | null;
};

export type ProductSupportChannel = {
  id: string;
  channel: string;
  label: string;
  target: string;
  availability?: string | null;
};

export type ProductFulfillmentSummary = {
  delivery?: ProductDeliveryEstimate;
  assurances?: ProductAssurancePoint[];
  support?: ProductSupportChannel[];
};

export type ProductMediaAsset = {
  id: string;
  clientId?: string | null;
  assetUrl: string;
  label?: string | null;
  storageKey?: string | null;
  usageTags?: ProductAssetUsageTag[] | null;
  altText?: string | null;
  displayOrder?: number | null;
  isPrimary?: boolean | null;
  checksum?: string | null;
  metadata?: Record<string, unknown> | null;
  createdAt?: string | null;
  updatedAt?: string | null;
};

export type ProductConfigurationPresetSelection = {
  optionSelections: Record<string, string[]>;
  addOnIds: string[];
  subscriptionPlanId?: string | null;
  customFieldValues: Record<string, string>;
};

export type ProductConfigurationPreset = {
  id: string;
  label: string;
  summary?: string | null;
  heroImageUrl?: string | null;
  badge?: string | null;
  priceHint?: string | null;
  displayOrder?: number | null;
  selection: ProductConfigurationPresetSelection;
};

export type ProductDetail = {
  id: string;
  slug: string;
  title: string;
  description?: string | null;
  category: string;
  basePrice: number;
  currency: string;
  status: "draft" | "active" | "archived";
  channelEligibility: string[];
  updatedAt?: string;
  optionGroups: ProductOptionGroup[];
  addOns: ProductAddOn[];
  customFields: ProductCustomField[];
  subscriptionPlans: ProductSubscriptionPlan[];
  fulfillmentSummary?: ProductFulfillmentSummary | null;
  mediaAssets?: ProductMediaAsset[];
  auditLog?: { id: string; action: string; createdAt: string }[];
  configurationPresets?: ProductConfigurationPreset[];
};
